daemon/cmd/kube_proxy_replacement.go

```diff
@@ -238,6 +238,30 @@ func initKubeProxyReplacementOptions() (strict bool) {
 		}
 	}
 
+	if !option.Config.EnableHostLegacyRouting {
+		switch {
+		case !option.Config.EnableNodePort:
+			fallthrough
+		case option.Config.Tunnel != option.TunnelDisabled:
+			fallthrough
+		// TODO: add multi-dev support
+		case len(option.Config.Devices) > 1:
+			fallthrough
+		// Non-BPF masquerade requires netfilter and hence CT.
+		case option.Config.Masquerade && !option.Config.EnableBPFMasquerade:
+			option.Config.EnableHostLegacyRouting = true
+		default:
+			found := false
+			if h := probesManager.GetHelpers("sched_cls"); h != nil {
+				_, found = h["bpf_redirect_neigh"]
+			}
+			if !found {
+				log.Warnf("Re-enabling host legacy routing as bypassing needs kernel 5.10.0 or newer.")
```

Comment:
Two things here:
1. We shouldn't warn on all released kernel versions.
1. There isn't any formatting directives here, this should be `log.Warn()`. If there were format directives, we should replace with structured logging (ie `log.WithFields(...).Warn(...)`.

Reply:
Makes sense. Added to: https://github.com/cilium/cilium/issues/13331
Could you elaborate wrt kernel versions? We should first be hitting the cases where we bail out w/o warning, so this should only show up on kube-proxy-free deployments (4.19+). Or do you think it would be better overall to just be silent here in general and then add info to the status dump? Maybe the latter is a better option?

Reply:
We should only warn when we are unable to implement some functionality the user explicitly asked for, typically if there's some serious suboptimal behaviour that we have to trigger to avoid getting into an even worse condition.

As of today, kernel v5.9-rc7 is available. This means there is at least another 2-3 months before the version specified in the log message is even released, let alone filtering down into distros or being supported by any of the regular suspects. It makes no sense to tell people to upgrade to a version that's not out yet. Even if v5.10.0 was out, the vast majority of users won't have the luxury of throwing out their current kernel and running the newer kernel. As-is, this code will add an extra warning message to a large proportion of deployments for potentially *years*. This makes users more likely to ignore all of the other warning messages, because we're always complaining about something that they can't do anything about anyway.

In this particular circumstance, we're talking about disabling an optimization. The actual forwarding behaviour should be correct whether the feature is enabled or not. Iff the user is highly sensitive to performance, then we can let them know via an Info message that upgrading to kernel 5.10.0 or later would improve performance.